utilities.py

```diff
@@ -0,0 +1,547 @@
+                ax = d.plot(kind='bar', title = chart_title, figsize=(20, 10), legend=False, fontsize=16)
+                ax
+            else:
+                print( "There is no data for this program and region after 2000." )
+
+        except AttributeError:
+            print("There's no data to chart for " + program.name + " !")
+
+
+#####################
+#  marker_text( row, is_echo ):
+#    
+# Create a string with information about the facility or program instance.
+#
+# Parameters:  
+#    row -- The row of data associated with the marker.
+#    is_echo -- If True, put some information with the marker to show the 
+#        programs that track the facility.
+#
+# Return:  The text to put with the marker
+#
+#####################
+
+def marker_text( row, is_echo ):
+    text = ""
+    if ( type( row['FAC_NAME'] == str )) :
+        try:
+            text = row["FAC_NAME"] + ' - '
+        except TypeError:
+            print( "A facility was found without name. ")
+        if ( is_echo ):
+            if ( row['AIR_FLAG'] == 'Y' ):
+                text += 'CAA, ' 
+            if ( row['NPDES_FLAG'] == 'Y' ):
+                text += 'CWA, ' 
+            if ( row['SDWIS_FLAG'] == 'Y' ):
+                text += 'SDWIS, ' 
+            if ( row['RCRA_FLAG'] == 'Y' ):
+                text += 'RCRA, ' 
+            if ( row['TRI_FLAG'] == 'Y' ):
+                text += 'TRI, ' 
+            if ( row['GHG_FLAG'] == 'Y' ):
+                text += 'GHG, ' 
+            text = text[:-1]
+            text += " - <p><a href='"+row["DFR_URL"]
+            text += "' target='_blank'>Link to ECHO detailed report</a></p>"
+    return text
+
+
+#####################
+#  mapper(df, is_echo=True):
+#    
+# Display a map of the DataFrame passed in.
+# Based on https://medium.com/@bobhaffner/folium-markerclusters-and-fastmarkerclusters-1e03b01cb7b1
+#
+# Parameters:  
+#    df -- A DataFrame containing records with latitude and longitude fields
+#        that can be plotted on a map
+#    is_echo -- A flag indicating whether the data contains ECHO_EXPORTER fields
+#        that can identify the EPA programs tracked for facilities
+#
+#####################
+
+def mapper(df, is_echo=True):
+    # Initialize the map
+    m = folium.Map(
+        location = [df.mean()["FAC_LAT"], df.mean()["FAC_LONG"]]
+    )
+
+    # Create the Marker Cluster array
+    #kwargs={"disableClusteringAtZoom": 10, "showCoverageOnHover": False}
+    mc = FastMarkerCluster("")
+ 
+    # Add a clickable marker for each facility
+    for index, row in df.iterrows():
+        mc.add_child(folium.CircleMarker(
+            location = [row["FAC_LAT"], row["FAC_LONG"]],
+            popup = marker_text( row, is_echo ),
+            radius = 8,
+            color = "black",
+            weight = 1,
+            fill_color = "orange",
+            fill_opacity= .4
+        ))
+    
+    m.add_child(mc)
+    bounds = m.get_bounds()
+    m.fit_bounds(bounds)
```

Comment:
Should this do any auto-zooming? When I do State/WA/Snohomish it shows me the whole United States
(It also shows a lot of stuff from outside the county, though not that far away – I assume that's data errors?)

Reply:
It should already be auto-zooming! That's what 
```
bounds = m.get_bounds()
m.fit_bounds(bounds)
```
is. But it may not be working because of the facilities EPA mis-geocoded. Though if they're not that far away it should at least zoom to them... Hmm.